### PR TITLE
[Performance, CPU] Rewriting OpenMP pragmas into parallel_for

### PR DESCRIPTION
**Pull request by tpatejko**
The PR replaces uses of OpenMP `parallel for` pragma with `parallel_for` primitive wherever it was possible (no `reduction`, `atomic`, `critical` or `barrier` clauses). 

This PR continues and includes work introduced by #3023 and #3027. The mentioned PRs contain initial benchmarking and information about performance regressions. This PR includes commits from earlier PRs and replaces additional pragmas. The earlier PR are going to be closed.

### CODE REVIEW

**Comment by tpatejko** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by tpatejko** [comment]
@VoVAllen could you trigger performance regression tests?

**Comment by tpatejko** [comment]
@VoVAllen @BarclayII could you trigger regression tests?

**Comment by VoVAllen** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by VoVAllen** [comment]
Sorry I just saw your comment

**Comment by BarclayII** [comment]
The largest speedup seems to be readout functions with 25%.

Here is the list of the benchmarks slowed down by 40%+.

| test_name | params | unit | number_master | machine | number_3171 | ratio |
|-|-|-|-|-|-|-|
| api.bench_edge_ids.track_time | 'cora', 'coo', 0.01, False | s | 0.000208 | c5.9xlarge-cpu | 0.000396 | 0.902062 |
| api.bench_in_degrees.track_time | 'cora', 'coo', 0.01 | s | 5.22E-05 | c5.9xlarge-cpu | 9.77E-05 | 0.87219 |
| api.bench_knn_graph.track_time | 10000, 32, 64, 'nn-descent' | s | 0.692811 | c5.9xlarge-cpu | 1.076581 | 0.553932 |
| api.bench_edge_ids.track_time | 'cora', 'coo', 0.1, False | s | 0.000462 | c5.9xlarge-cpu | 0.000689 | 0.491037 |
| api.bench_knn_graph.track_time | 10000, 32, 8, 'nn-descent' | s | 0.072226 | c5.9xlarge-cpu | 0.104481 | 0.446579 |
| api.bench_builtin_apply_edges.track_time | 'cora', 'coo', 512, 'u+v' | s | 0.0013 | c5.9xlarge-cpu | 0.001876 | 0.442883 |
| api.bench_knn_graph.track_time | 1000, 4, 8, 'nn-descent' | s | 0.007045 | c5.9xlarge-cpu | 0.010088 | 0.431965 |
| api.bench_knn_graph.track_time | 10000, 4, 8, 'nn-descent' | s | 0.083871 | c5.9xlarge-cpu | 0.118583 | 0.413867 |

**Comment by tpatejko** [comment]
@VoVAllen could you trigger performance regression tests?

**Comment by VoVAllen** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by BarclayII** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by BarclayII** [comment]
The benchmark result goes as follows, with some of the `update_all` execution on CSC gets 30% slower (mostly `u_mul_e` message function and `sum` reduce function on small feature sizes)

  | test_name | params | unit | number_3171 | machine | number_master | ratio
-- | -- | -- | -- | -- | -- | -- | --
8 | api.bench_builtin_apply_edges.track_time | 'cora', 'coo', 128, 'u->e' | s | 0.000407 | c5.9xlarge-cpu | 0.000195 | 1.086563
197 | api.bench_homograph_scipy_construction.track_time | 'small', 'coo' | s | 0.000137 | c5.9xlarge-cpu | 6.81E-05 | 1.009174
9 | api.bench_builtin_apply_edges.track_time | 'cora', 'coo', 128, 'u+v' | s | 0.000473 | c5.9xlarge-cpu | 0.000247 | 0.912796
16 | api.bench_builtin_apply_edges.track_time | 'cora', 'csr', 512, 'u->e' | s | 0.001269 | c5.9xlarge-cpu | 0.000767 | 0.655484
374 | api.bench_random_walk.track_time | 'livejournal', 1000, 20, '_node2vec' | s | 0.001473 | c5.9xlarge-cpu | 0.000946 | 0.557238
69 | api.bench_builtin_update_all_csc.track_time | 'ogbn-arxiv', 'csc', 4, 'u_mul_e', 'sum' | s | 0.002903 | c5.9xlarge-cpu | 0.002016 | 0.440368
237 | api.bench_knn_graph.track_time | 1000, 4, 8, 'bruteforce-blas' | s | 0.00328 | c5.9xlarge-cpu | 0.002344 | 0.398937
11 | api.bench_builtin_apply_edges.track_time | 'cora', 'coo', 512, 'u+v' | s | 0.001333 | c5.9xlarge-cpu | 0.000972 | 0.372171
427 | api.bench_readout.track_time | 256, 16, 'min', 'edge' | s | 0.000238 | c5.9xlarge-cpu | 0.000177 | 0.344443
88 | api.bench_builtin_update_all_csc.track_time | 'reddit', 'csc', 4, 'u_mul_e', 'mean' | s | 0.39102 | c5.9xlarge-cpu | 0.291074 | 0.343371
87 | api.bench_builtin_update_all_csc.track_time | 'reddit', 'csc', 4, 'u_mul_e', 'sum' | s | 0.389416 | c5.9xlarge-cpu | 0.289952 | 0.343036
106 | api.bench_builtin_update_all_csc.track_time | 'ogbn-proteins', 'csc', 4, 'u_mul_e', 'mean' | s | 0.269016 | c5.9xlarge-cpu | 0.200894 | 0.339094
70 | api.bench_builtin_update_all_csc.track_time | 'ogbn-arxiv', 'csc', 4, 'u_mul_e', 'mean' | s | 0.00373 | c5.9xlarge-cpu | 0.0028 | 0.331844
105 | api.bench_builtin_update_all_csc.track_time | 'ogbn-proteins', 'csc', 4, 'u_mul_e', 'sum' | s | 0.268104 | c5.9xlarge-cpu | 0.2016 | 0.329882
229 | api.bench_in_subgraph.track_time | 'reddit', 'csc', 5000 | s | 0.023506 | c5.9xlarge-cpu | 0.017932 | 0.310838
125 | api.bench_edge_ids.track_time | 'cora', 'csr', 0.01, False | s | 0.000227 | c5.9xlarge-cpu | 0.000174 | 0.303494

**Comment by tpatejko** [comment]
@BarclayII Could you trigger performance tests?

I changed the `parallel_for` implementation in order to limit number of call to task lambdas. For the performance tests with small data sizes repeated function calls in the loop could result in additional overhead and performance regression.

In this implementation, `parallel_for` requires a lambda that takes two parameters, beginning and end of the range. There's also an additional for-loop in the lambda's body.

**Comment by BarclayII** [comment]
@dgl-bot run c5.9xlarge kernel,api dmlc/master

**Comment by tpatejko** [comment]
@BarclayII I see that there are still some regressions. I collected the results for cases with 20% slowdown.
I also executed the performance tests on our internal CI. Most of the cases are confirmed, with exception for some `readout` cases, `sample_neighbors` and `udf_update_all` cases.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/tpatejko/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/tpatejko/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  |   |   |   | Github |   |   | Internal |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   |   |   | Master | PR | Ratio | Master | PR | Ratio
69 | api.bench_builtin_update_all_csc.track_time | 'ogbn-arxiv', 'csc', 4, 'u_mul_e', 'sum' | s | 0.002142 | 0.002978 | 39.00667 | 0.002098 | 0.00249 | 18.6674
70 | api.bench_builtin_update_all_csc.track_time | 'ogbn-arxiv', 'csc', 4, 'u_mul_e', 'mean' | s | 0.002872 | 0.003724 | 29.65746 | 0.002831 | 0.003485 | 23.11899
87 | api.bench_builtin_update_all_csc.track_time | 'reddit', 'csc', 4, 'u_mul_e', 'sum' | s | 0.288187 | 0.389302 | 35.08679 | 0.251854 | 0.300917 | 19.48072
88 | api.bench_builtin_update_all_csc.track_time | 'reddit', 'csc', 4, 'u_mul_e', 'mean' | s | 0.290025 | 0.394061 | 35.87135 | 0.25385 | 0.302694 | 19.24125
105 | api.bench_builtin_update_all_csc.track_time | 'ogbn-proteins', 'csc', 4, 'u_mul_e', 'sum' | s | 0.203356 | 0.264293 | 29.96567 | 0.191394 | 0.220862 | 15.39679
106 | api.bench_builtin_update_all_csc.track_time | 'ogbn-proteins', 'csc', 4, 'u_mul_e', 'mean' | s | 0.203602 | 0.266956 | 31.11676 | 0.191018 | 0.220678 | 15.52727
129 | api.bench_edge_ids.track_time | 'cora', 'csc', 0.01, False | s | 0.000189 | 0.000265 | 39.99919 | 0.000252 | 0.000298 | 18.0397
262 | api.bench_knn_graph.track_time | 10000, 4, 8, 'bruteforce' | s | 0.040429 | 0.050585 | 25.12037 | 0.050068 | 0.055993 | 11.8345
431 | api.bench_readout.track_time | 256, 128, 'sum', 'edge' | s | 0.000343 | 0.000476 | 38.87093 | 0.000243 | 0.00037 | 52.14224
437 | api.bench_readout.track_time | 256, 128, 'mean', 'edge' | s | 0.000397 | 0.000526 | 32.47723 | 0.000335 | 0.000376 | 12.08825
439 | api.bench_readout.track_time | 256, 512, 'sum', 'edge' | s | 0.001546 | 0.001959 | 26.69503 | 0.001155 | 0.00117 | 1.335369
445 | api.bench_readout.track_time | 256, 512, 'mean', 'edge' | s | 0.001618 | 0.002037 | 25.84721 | 0.001332 | 0.001304 | -2.06112
514 | api.bench_sample_neighbors.track_time | 'reddit', 'csc', 20000, 20 | s | 0.004243 | 0.005482 | 29.22649 | 0.003689 | 0.003499 | -5.14895
574 | api.bench_udf_update_all.track_time | 'pubmed', 'coo', 8, 'copy_u', 'sum' | s | 0.134991 | 0.163856 | 21.38245 | 0.150958 | 0.148873 | -1.38128



</body>

</html>


I'm not quite sure how to proceed with the PR. The implementation in the PR of the `parallel_for` primitive is similar to the one in Pytorch: the interface and the implementation are changed in order to reduce the number of function calls to the task per thread.

We also executed training with the changes in the PR and no performance regressions were encountered:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/tpatejko/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/tpatejko/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

<div style='box-sizing: border-box;font-variant-ligatures: normal;font-variant-caps: normal;
orphans: 2;text-align:start;widows: 2'>



model | dataset | master | PR | speedup
-- | -- | -- | -- | --
gat_ns | ogbn-products | 208.451482 | 208.422 | 1
gat_ns | reddit | 73.361161 | 65.3351 | 1.12
rgcn_hetero_ns | am | 5.270454 | 2.58049 | 2.04
rgcn_hetero_ns | ogbn-mag | 142.622971 | 65.5553 | 2.18
rgcn_homogeneous_ns | am | 1.532482 | 1.48914 | 1.03
rgcn_homogeneous_ns | ogbn-mag | 268.887767 | 276.848 | 0.97
sage_ns | ogbn-products | 63.54111 | 52.0431 | 1.22
sage_ns | reddit | 66.257414 | 66.3105 | 1



</body>

</html>

The tests were conducted the following way:
dry run: 3 epochs
timing: 10 epochs
 
GAT
hidden feats: 120
num heads: 4
num layers: 3
dropout: 0.5
fanouts: 10, 10, 10
batch size: 6554
 
RGCN
hidden feats: 64
num bases: 2
num layers: 2
dropout: 0.5
self loop: True
fanouts: 25, 15
batch size: 1024
 
GraphSAGE

hidden feats: 256
num layers: 3
dropout: 0.5
fanouts: 5, 10, 15
batch size: 1000

**Comment by jermainewang** [comment]
Just a guess. Is the regression caused by too many invocation of indirect functions?

**Comment by jermainewang** [comment]
Since the regression happens in the below-millisecond regime, I agree that the problem is mainly due to repeated function calls. Given that the end-to-end results look promising and having a better OpenMP control has been a long-wished feature, I will say yes to this change. @tpatejko Please help pass the CI and resolve the conflict. Thanks a lot for the effort!

**Review by BarclayII — APPROVED** [review]

**Comment by erickim555** [comment]
Hello, this PR may be making `COOToCSR()` use ~3-5x more CPU memory than before. In https://github.com/dmlc/dgl/issues/3307, before this PR I found that it was necessary for me to set `OMP_NUM_THREADS=10` for `COOToCSR()` to succeed (eg not CPU OOM) on my large graph (billions of nodes, edges).
But now, with this PR I have to decrease `OMP_NUM_THREADS` to a smaller value, otherwise I get a CPU OOM. With `OMP_NUM_THREADS=2`, I'm seeing peak memory usage of ~36% of my available CPU memory during `COOToCSR()`, whereas before I could bump up to `OMP_NUM_THREADS=10` and things would still work.

Worth noting that this work-in-progress PR may resolve this memory issue: https://github.com/dmlc/dgl/pull/3326/files

**Comment by erickim555** [comment]
**UPDATE**: oops, my mistake, my env wasn't set up correctly. I can confirm that my setup successfully runs with this code change without CPU OOM-ing, feel free to ignore my above comment. Sorry about the confusion!
